File: continuation.ts
type ContFunc<R, A> = (k: (a: A) => R) => R;

class Cont<R, A> {
  constructor(public readonly runCont: ContFunc<R, A>) {}

  bind<B>(f: (a: A) => Cont<R, B>): Cont<R, B> {
    return new Cont((k) => this.runCont((a) => f(a).runCont(k)));
  }

  flatMap<B>(f: (a: A) => Cont<R, B>): Cont<R, B> {
    return new Cont((k) => this.runCont((a) => f(a).runCont(k)));
  }

  static of<R, A>(a: A): Cont<R, A> {
    return new Cont((k) => k(a));
  }

  map<B>(f: (a: A) => B): Cont<R, B> {
    return this.bind((a) => Cont.of<R, B>(f(a)));
  }

  ap<B>(fab: Cont<R, (a: A) => B>): Cont<R, B> {
    return fab.bind((f) => this.map(f));
  }
}

const example = Cont.of<string, string>("Hello").bind((x) =>
  Cont.of<string, string>("World").map((y) => `${x} ${y}`)
);

const result = example.runCont((s) => s);
console.log("Result:", result);

function callCC<R, A, B>(
  f: (exit: (a: A) => Cont<R, B>) => Cont<R, A>
): Cont<R, A> {
  return new Cont<R, A>((k: (a: A) => R) =>
    f((a: A) => new Cont<R, B>((_ignore: (b: B) => R) => k(a))).runCont(k)
  );
}

const computation = Cont.of<number, number>(10)
  .flatMap((x) => Cont.of<number, number>(x * 2))
  .flatMap((x) => Cont.of<number, number>(x + 5));

const computationWithExit = callCC<number, number, number>((exit) =>
  Cont.of<number, number>(10).flatMap((x) =>
    x > 5 ? exit(999) : Cont.of<number, number>(x * 2)
  )
).flatMap((x) => Cont.of<number, number>(x + 1));

const result3 = computationWithExit.runCont((x) => x);
console.log(result3);

File: continuation.hs
{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
{-# HLINT ignore "Eta reduce" #-}
module Main where

newtype Cont r a = Cont { runCont :: (a -> r) -> r }

instance Functor (Cont r) where
  fmap f (Cont k) = Cont (\br -> k (br . f))

instance Applicative (Cont r) where
  pure a = Cont (\ar -> ar a)
  (Cont kf) <*> (Cont ka) = Cont (\br -> kf (\f -> ka (br . f)))

instance Monad (Cont r) where
  (Cont ka) >>= kab = Cont (\br -> ka (\a -> runCont (kab a) br))
  return = pure

example :: Cont r String
example = do
  x <- Cont $ \k -> k "Hello"
  y <- Cont $ \k -> k "World"
  return (x ++ " " ++ y)

main :: IO ()
main = putStrLn $ runCont example id
